Add API to modify pathway

twitu opened the pull request:
Wiki entry: https://github.com/dush-t/pathways/wiki/Schema-design-for-modifying-a-pathway

twitu committed:
Add API to modify pathway

Wiki entry: https://github.com/dush-t/pathways/wiki/Schema-design-for-modifying-a-pathway

dusnat commented on server/src/db/queries/pathway/modifyPathwaySteps.js:
```diff
@@ -0,0 +1,119 @@
+    })
+
+    console.log("Deleting steps")
+    console.log(steps)
```
Remove the extra console.log statements

dusnat commented on server/src/db/queries/pathway/modifyPathwaySteps.js:
```diff
@@ -0,0 +1,119 @@
+    })
+
+    console.log("Updating steps")
+    console.log(steps)
```
Remove the additional console log statements

dusnat commented on server/src/db/queries/pathway/modifyPathwaySteps.js:
```diff
@@ -0,0 +1,119 @@
+    })
+
+    console.log("Creating new steps ran")
+    console.log(props)
```
Remove the additional console.log statements

dusnat commented on server/src/db/queries/pathway/modifyPathwaySteps.js:
```diff
@@ -0,0 +1,119 @@
+
+const updatePathwaySteps = ({id, updateSteps}) => {
+    const steps = []
+    updateSteps.forEach((step) => {
```
Instead of using a `forEach` here, maybe use the `map` method, since you're only creating one new array from an existing one.

dusnat commented on server/src/db/queries/pathway/modifyPathwaySteps.js:
```diff
@@ -0,0 +1,119 @@
+
+const deletePathwaySteps = ({id, deleteSteps}) => {
+    const steps = []
+    deleteSteps.forEach((step) => {
```
Again, since you're only converting one array into another, I think you should use `map` instead of `forEach`

dusnat commented on server/src/db/queries/pathway/modifyPathwaySteps.js:
```diff
@@ -0,0 +1,119 @@
+        UNWIND $steps as upd
+        WITH p, upd
+        MATCH (p)<-[:HAS_PARENT_PATHWAY]-(s:Step {id: upd.id})
+        SET s.name = upd.name, s.difficulty = upd.difficulty, s.index = upd.index
```
This query does not allow updating the step content, even though it's a part of `StepPayload`. Are you going to write a different endpoint for updating content?

twitu replied:
Yes I am assuming three layers,
1. Raw data - md file - step payload
2. Step (can be pathway) - meta data - time complete
3. Pathway (made of steps) - additional meta data - order of steps

So each layers should have different endpoints.

dusnat replied:
I have a problem with this approach. I see no reason why these layers should be decoupled for the client too. For the client, I think that a step should be an atomic entity. The client does not need to know how we're splitting a step into metadata and content.

twitu replied:
Let me try to explain my position better. I feel that here there are two types of data here,
1. The actual instructions to learn something
2. The interlinking between them using the concept of steps and pathways

These should necessarily be decoupled in UX as well as the APIs. The reason being that a user will interact with them separately. For the actual instructions an author will probably make a PR or edit it live. But this will only change the instructions to learn something, we can call it `Resource`. For the linking part, an author will create a step pointing to a resource (or an existing pathway) and link it to a new pathway. Here the author will probably use a graph view that allows them to easily search and link.

Another reason I feel these two types of data should have separate interfaces is to ensure the atomicity of the `Resource`. That is a `Resource` should exist standalone. It should not be created to suit the needs of one particular pathway. With this we can ensure that authors follow a certain flow:
1. Create a new `Resource` for a topic
2. See if a pathway for a particular skill can be created using existing `Resource` or pathways

dusnat commented on server/src/db/queries/pathway/modifyPathwaySteps.js:
```diff
@@ -0,0 +1,119 @@
+        WITH p
+        UNWIND $steps as del
+        WITH p, del
+        MATCH (p)<-[:HAS_PARENT_PATHWAY]-(s:Step{id: del.id})
```
This query will delete the node that contains the metadata, but in case that node does not point to a pathway, it will not delete the node containing the content. Not desired. You should delete the content node too if the Step node is not pointing to a pathway.

twitu replied:
Again going by the three layer style, the actual content of the step has only one instance i.e. an md file or some other file storage. This step is just a pointer to it with some meta data, deleting the pointer should suffice.

dusnat reviewed (changes requested):
Some changes are going to make these queries significantly more complex. I suggest having only one possible query for one mutation (instead of three, like in this case).

So I suggest having three mutations (for create, update and delete) instead of just one, and put the three queries in different files. This would also make it easier for the front-end developer to keep better track of what he's doing with the API.

twitu committed:
Improve API for pathways CRUD

Make two apis:
1. createUpdatePathway: create or modify pathway and steps
2. deletePathway: delete entire pathway or steps
